### DISCUSSION TITLE
Use case of "save_to_file" in TM1 services

### DISCUSSION THREAD

**Discussion by macsir** [cubewise-code/tm1py#822] [General]
Hi, @MariusWirtz  Could you elaborate some use cases of this function to avoid multiple authentication with TM1? Thanks.

**Comment by MariusWirtz** [comment]
Yes. Imagine this scenario: 
- CAM Authentication that takes a couple of seconds
- TM1py is called regularly from users through a TI process (think of a process button in tm1web or something)

If TM1py can perform the task in no time, it would be unpleasant and unreasonable for the users to wait 3-5 seconds for the CAM authentication.

In that scenario, it can be useful to persist your session on the server, so that it can be re-used for multiple runs.

Of course, once the session times out you have to re-authenticate.

``` python
import time

from TM1py import TM1Service

tm1 = TM1Service(address="", port=12354, ssl=True, user="admin", password="apple")

tm1.save_to_file("tm1conn")

time.sleep(5)

tm1: TM1Service = TM1Service.restore_from_file("tm1conn")
# session timeout
if not tm1.connection.is_connected():
    tm1.re_authenticate()

print(tm1.server.get_server_name())


```

**Reply by macsir** [reply]
Thanks for the prompt reply. @MariusWirtz  Yes, that does make sense and the timeout value I guess is based on HTTPSessionTimeoutMinutes in tm1s.cfg file.

**Reply by MariusWirtz** [reply]
Correct!

**Reply by macsir** [reply]
@MariusWirtz  two questions, could you help on those? Thanks.
1. I created a saved tm1 session and tried to re-use it after 4 hours (HTTPSessionTimeoutMinutes=240). It still works and can log into TM1 and execute something?
2. I tried to run `re_authenticate()` directly even if not timed out yet. It gives me the following errors. my TM1py is up-to-date.
```
    tm1.re_authenticate()
  File "/Users/mmorgs/Library/Python/3.9/lib/python/site-packages/TM1py/Services/TM1Service.py", line 86, in re_authenticate
    self._tm1_rest = RestService(**self.connection._kwargs)
AttributeError: 'RestService' object has no attribute '_kwargs'
```

**Reply by macsir** [reply]
1. Tested again this morning, it really timed out. However, I am not sure when TM1 starts to count the starting time of a session. I assume it starts when I created this saved session using the method.
2. Still have this error when trying to re-auth the timed-out saved session.
I will keep updated if I find anything. Thanks.

**Reply by MariusWirtz** [reply]
I can't replicate the error in the `re_authenticate` function. Does this work?

``` python
from TM1py import TM1Service

tm1 = TM1Service(address="", port=12354, ssl=True, user="admin", password="apple")
tm1.logout()
tm1.re_authenticate()
server_name = tm1.server.get_server_name()

print(server_name)

```

**Reply by macsir** [reply]
Thanks, @MariusWirtz  your code works for me. Looks like I need to logout explicitly after creating a saved session so that I can re auth again from a timeout saved session. Something like that
```python
import time

from TM1py import TM1Service

tm1 = TM1Service(address="", port=12354, ssl=True, user="admin", password="apple")

tm1.save_to_file("tm1conn")
tm1.logout() #explicit logout

time.sleep(5)

tm1: TM1Service = TM1Service.restore_from_file("tm1conn")
# session timeout
if not tm1.connection.is_connected():
    tm1.re_authenticate()

print(tm1.server.get_server_name())
```
I will keep testing.

**Reply by macsir** [reply]
Hi, @MariusWirtz  Sorry for late reply and got some time to test today. As Ryan pointed out, the server works fine without explicit logout. For some reason, I messed up my previous testing. Thanks.

**Comment by rclapp** [comment]
If you logout you will never be able to use a saved session.

Sent from my mobile device

On Nov 3, 2022 12:45 AM, tm1sir ***@***.***> wrote:

Thanks, @MariusWirtz<https://github.com/MariusWirtz> your code works for me. Looks like I need to logout explicitly after creating a saved session so that I can re auth again from a timeout saved session. Something like that

import time

from TM1py import TM1Service

tm1 = TM1Service(address="", port=12354, ssl=True, user="admin", password="apple")

tm1.save_to_file("tm1conn")
**tm1.logout()**

time.sleep(5)

tm1: TM1Service = TM1Service.restore_from_file("tm1conn")
# session timeout
if not tm1.connection.is_connected():
    tm1.re_authenticate()

print(tm1.server.get_server_name())

I will keep testing.

**Reply by macsir** [reply]
Yes, you are right, @rclapp  Thanks.

**Comment by rkvinoth** [comment]
Just to add some more context, In some TM1py versions the pickled file had the credentials stored in plain format. 
So, you gotta be more careful on where you are going to save the file.

**Reply by MariusWirtz** [reply]
Good points @rkvinoth.
It depends on the authentication mode too. password is only stored in the file if a password-based TM1 login is used. 
If you are using CAM SSO or Integrated you wouldn't face this issue.

If you don't want the password to be persisted, you can drop it from the TM1Service object like.
Of course, that means you can't call the `re_authenticate` function later on.

``` python
from TM1py import TM1Service

tm1 = TM1Service(address="", port=12354, ssl=True, user="admin", password="apple")

del tm1.connection._kwargs["password"]
tm1.save_to_file("C:/temp/conn")

tm1 = TM1Service.restore_from_file("C:/temp/conn")

print(tm1.server.get_server_name())
```

**Reply by macsir** [reply]
Yes, agreed. Thanks, guys.
